mev-boost-rs/src/relay_mux.rs

```diff
@@ -143,11 +145,40 @@ impl BlindedBlockProvider for RelayMux {
         &self,
         bid_request: &BidRequest,
     ) -> Result<SignedBuilderBid, BlindedBlockProviderError> {
-        let responses = stream::iter(self.relays.iter().cloned())
+        let time_out =
+            tokio::time::sleep(tokio::time::Duration::from_secs(FETCH_BEST_BID_TIME_OUT));
+        tokio::pin!(time_out);
+
+        let mut requests = stream::iter(self.relays.iter().cloned())
             .map(|relay| async move { relay.fetch_best_bid(bid_request).await })
-            .buffer_unordered(self.relays.len())
-            .collect::<Vec<_>>()
-            .await;
+            .buffer_unordered(self.relays.len());
+
+        let mut responses = Vec::new();
+
+        loop {
+            tokio::select! {
+                res = requests.next() => {
+                    match res {
+                        Some(bid) => {
+                            responses.push(bid);
+                        }
+                        None => {
+                            break;
+                        }
+                    }
+                }
+                _ = &mut time_out => {
+                    break;
+                }
+            }
+        }
```

Comment:
nice! I'll suggest a bit different implementation and provide rationale for why I prefer it

1. rather than make a `time_out` that we manually wait for in the `select!` macro, I think we can be more concise with the `tokio::time::timeout` helper: https://docs.rs/tokio/latest/tokio/time/fn.timeout.html (and based on how this function is written, we can also us a `std` `Duration` rather than the `tokio` specific one
2. the next point concerns how to apply the timeout: does it timeout the full collection (like you have here), _or_ should we write a timeout per `fetch_best_bid` call? and I think we should use the latter, as what you have written I am pretty sure will "leak" some futures (if you look at the `timeout` docs, they explicitly say they cancel the future upon timeout and I think what you have will keep the futures hanging in `requests` and proceed -- over time, this would lead to leaking memory)

Reply:
How about calling `drop` right after the loop and passing the futures in requests to it. Then we don't hold the futures till the end of the scope of this function. Will checkout tokio's timeout :)

Reply:
you could probably `drop` `requests` but I consider this a bit un-idiomatic given "well-known" primitives for exactly this type of task in `tokio`, namely this `timeout` fn

otherwise we are kind of just reimplementing timeout functionality when it is allegedly well-tested, well-optimized etc in the library implementation